embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/FailedApiCallsPerEndpoint.kt

```diff
@@ -0,0 +1,80 @@
+package io.embrace.android.embracesdk.comms.delivery
+
+import io.embrace.android.embracesdk.comms.api.EmbraceApiService.Companion.Endpoint
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * A map containing a list of failed API calls per endpoint.
+ */
+internal class FailedApiCallsPerEndpoint {
+
+    private val failedApiCallsMap = ConcurrentHashMap<Endpoint, DeliveryFailedApiCalls>()
+
+    /**
+     * Adds a failed API call in the corresponding endpoint's list.
+     */
+    fun add(endpoint: Endpoint, failedApiCall: DeliveryFailedApiCall) {
+        val failedApiCallsForEndpoint = failedApiCallsMap.getOrPut(endpoint) { DeliveryFailedApiCalls() }
+        failedApiCallsForEndpoint.add(failedApiCall)
+    }
+
+    /**
+     * Returns the list of failed API calls for the corresponding endpoint.
+     */
+    fun get(endpoint: Endpoint): DeliveryFailedApiCalls? {
+        return failedApiCallsMap[endpoint]
+    }
+
+    /**
+     * Clears all lists of failed API calls.
+     */
+    fun clear() {
+        failedApiCallsMap.clear()
+    }
+
+    /**
+     * Returns the total number of failed API calls in all endpoints' lists.
+     */
+    fun failedApiCallsCount() = failedApiCallsMap.values.sumOf { it.size }
+
+    /**
+     * Returns the number of failed API calls in the corresponding endpoint's list.
+     */
+    fun failedApiCallsCount(endpoint: Endpoint) = failedApiCallsMap[endpoint]?.size ?: 0
+
+    /**
+     * Returns true if there are any failed API calls in any endpoint's list.
+     */
+    fun hasAnyFailedApiCalls(): Boolean {
+        return failedApiCallsMap.values.any { it.isNotEmpty() }
+    }
+
+    /**
+     * Returns true if there are no failed API calls in any endpoint's list.
+     */
+    fun hasNoFailedApiCalls(): Boolean {
+        return !hasAnyFailedApiCalls()
+    }
+
+    /**
+     * Returns the next failed API call to be sent and removes it from the corresponding
+     * endpoint's list.
+     */
+    fun pollNextFailedApiCall(): DeliveryFailedApiCall? {
```

Comment:
probably need to synchronize this call with any methods that can modify the internal map? Given the amount of traffic we have, I don't think we need to be more clever than that the the locking

Reply:
As we talked on Slack, this class is accessed by a single thread, so I just added a comment in case we start accessing it from multiple threads.